### PR TITLE
Update to the latest version of argparse

### PR DESCRIPTION
**Pull request by adamjstewart**
I just noticed that the vendored version of argparse we were using (1.2.1) is from 2014-05-09. There is a new release available (1.4.0) from 2015-09-12 which should be stable.

### CODE REVIEW

File: lib/spack/external/argparse.py [hunk 1]
```diff
@@ -108,8 +94,6 @@
 import sys as _sys
 import textwrap as _textwrap
 
-from llnl.util.tty.colify import colified
```
Comment (adamjstewart) [inline lib/spack/external/argparse.py]: It looks like we had slightly modified the vendored copy of argparse. Is this still needed? In general, I'm not too keen on modifying vendored packages, as it makes it more difficult to upgrade.

Comment (adamjstewart) [reply lib/spack/external/argparse.py]: Actually, it looks like we've modified argparse several times:
https://github.com/spack/spack/commits/develop/lib/spack/external/argparse.py

@tgamblin Can you comment?

**Comment by tgamblin** [comment]
Yep that was me.  Here's the history:

```console
$ git lg argparse.py
* 21fae63 - Add Python 3 aliases to our argparse backport. (2 years ago) <Todd Gamblin>
* a5859b0 - Add ability to get subparser by name from argparse (3 years, 2 months ago) <Todd Gamblin>
* a8ed1ec - Minor argparse improvement. (3 years, 3 months ago) <Todd Gamblin>
* 7714d08 - Remvoe dependence on v2.7 argparse by including argparse. (3 years, 5 months ago) <Todd Gamblin>
```

`21fae63` adds aliases to the backport, but it looks like `1.4.0` includes that, so that is not needed.  `a5859b0` added a function to get subparsers by name, but I can't find where we use that anymore.  I think it can go.

The last one is `a8ed1ec`, which makes one significant change:

```diff
@@ -2285,8 +2287,8 @@ class ArgumentParser(_AttributeHolder, _ActionsContainer):
     def _check_value(self, action, value):
         # converted value must be one of the choices (if specified)
         if action.choices is not None and value not in action.choices:
-            tup = value, ', '.join(map(repr, action.choices))
-            msg = _('invalid choice: %r (choose from %s)') % tup
+            cols = colified(sorted(action.choices), indent=4, tty=True)
+            msg = _('invalid choice: %r choose from:\n%s') % (value, cols)
             raise ArgumentError(action, msg)
```

It makes the output better when you make an invalid choice for an option.  So you get this:

```
$ spack config --scope badscope
usage: spack config [-h]
                    [--scope {defaults,defaults/darwin,system,system/darwin,site,site/darwin,user,user/darwin}]
                    SUBCOMMAND ...
spack config: error: argument --scope: invalid choice: 'badscope' choose from:
    defaults         site         system         user
    defaults/darwin  site/darwin  system/darwin  user/darwin
```

instead of this:

```
$ spack config --scope badscope
usage: spack config [-h]
                    [--scope {defaults,defaults/darwin,system,system/darwin,site,site/darwin,user,user/darwin}]
                    SUBCOMMAND ...
spack config: error: argument --scope: invalid choice: 'badscope' (choose from 'defaults', 'defaults/darwin', 'system', 'system/darwin', 'site', 'site/darwin', 'user', 'user/darwin')
```

I rather like that one, and it's like 2 lines so I don't think we need to get rid of it.

**Comment by adamjstewart** [comment]
@tgamblin I'll re-add the `colified` logic. I'm not super keen on modifying external packages, but perhaps we can try to submit a PR with `argparse` to support this directly someday.

**Comment by tgamblin** [comment]
If you want it separate, you could just override _check_value in spack.main.SpackArgumentParser, and then the external code doesn’t need to be touched.

**Comment by tgamblin** [comment]
Looked at the code again and you can't really override the inner `_check_value` function without overriding a huge amount of code.  So upstreaming seems like the best option.

**Pull request merged**